### ISSUE TITLE
Methods for reordering nodes and individuals

### ISSUE THREAD

**Issue by hyanwong** [tskit-dev/tskit#2060]
I would like to be able to reorder the individuals so that the IDs change but the structure is all the same. I would similarly like to be able to do this for nodes. I don't think there are built-in functions to do this yet, although `TreeSequence.subset` will do it as a side effect for nodes, if given an ordering that lists all the node IDs.

**Comment by petrelharp** [comment]
Ok - so, what exactly do you need to do? `subset( )` will let you put the nodes in arbitrary order, and then orders individuals by first occurrence of node, IIRC.

**Comment by benjeffery** [comment]
So I'm assuming this would be a method on `TableCollection` that would take an array for each table that specifies the new order (or `None` for no change)? Sounds do-able. I'd like to understand a little more about the context you're doing this in though?

**Comment by hyanwong** [comment]
Actually it turns out that I don't need it in this instance. But there have been times in the past when I have wanted to do this.

Here, I wanted it to be able to order all individuals at a given time (in generations) so that when visualising their pedigree, the lines connecting them had crossings minimised. I had the order within each generation, but it was fiddly sorting the indexes in the right way.

Previously I've wanted, for example, to put all the sample nodes as the first n nodes, even when some are historical samples (and so there might be younger internal nodes). But that's just off the top of my head.

**Comment by hyanwong** [comment]
Hi @benjeffery - this seems like it would be useful in some tsinfer improvements that @szhan might make (see https://github.com/tskit-dev/tsinfer/issues/189#issuecomment-1032042549 for context). I think this might be a simple enough addition to tskit for him to try making a PR about it. Does that seem sensible? Are there any misgivings about adding the functionality to tskit?

**Comment by hyanwong** [comment]
I guess at a minimum we would want `TableCollection.reorder_tables(nodes=X, individuals=Y, populations=Z)`, since those are the only tables which are allowed to have an arbitrary order. I'm not sure about migrations?

The ability to reorder the other tables edges, sites, etc to form a non-valid table collection is a bit esoteric, so not sure we want to bother with that.

**Comment by petrelharp** [comment]
How will this interact with `subset`? For instance, if `nodes=X` would `X` be required to have *all* of the nodes? One way forward would be to add `individual_order=Y` and `population_order=Z` arguments to `subset( )` (and then possibly rename it `reorder( )` since maybe that's a better name)?

> The ability to reorder the other tables edges, sites, etc to form a non-valid table collection is a bit esoteric, so not sure we want to bother with that.

Definitely.

**Comment by benjeffery** [comment]
> Are there any misgivings about adding the functionality to tskit?

I don't think so - is this a perf sensitive application? We normally add these things in at the python layer, at least initially and only push down to C if needed. @szhan should go ahead and PR something for sure, especially if it isn't much work to get to a prototype that we can concretely discuss.

**Comment by hyanwong** [comment]
No, I don't think this is perf-sensitive. It might be done for a large tree sequence, but it is only done once in the entire pipeline, so it will be a minuscule part of the whole.

I guess we want to decide on whether as @petrelharp says we want to roll this into the "subset" method?

**Comment by benjeffery** [comment]
> However, I guess we want to decide on whether (as @petrelharp says) we want to roll this into the "subset" method?

I'd rather have a separate method so the two uses are clear, even if there is some small overlap in functionality.

**Comment by hyanwong** [comment]
OK, If @petrelharp agrees, I think the method / API is well defined enough for @szhan to have a go at it:

```
TableCollection.reorder(nodes=None, individuals=None, populations=None)
```

Acts in place, and could potentially return a tuple of the 3 maps of old->new indexes?

**Comment by petrelharp** [comment]
Soiunds good to me!

**Comment by hyanwong** [comment]
Here's some demo code showing the logic, @szhan. I *think* I have it right, but check that the mapping is the right way around.

```
import numpy as np
import msprime
ts = msprime.sim_ancestry(10, sequence_length=100, recombination_rate=0.1, random_seed=1)
ts = msprime.sim_mutations(ts, rate=0.1, random_seed=1)

def reorder_nodes(tables, node_map):
    tables.nodes.replace_with(tables.nodes[node_map])
    tables.edges.parent = node_map[tables.edges.parent]
    tables.edges.child = node_map[tables.edges.child]
    tables.mutations.node = node_map[tables.mutations.node]
    tables.sort()


reversed_nodes = np.arange(ts.num_nodes)[::-1].astype(tables.edges.parent.dtype)
table_collection = ts.dump_tables()
reorder_nodes(table_collection, reversed_nodes)
new_ts = table_collection.tree_sequence()
# check the diversity is the same
print(ts.diversity(), new_ts.diversity())
```

**Comment by benjeffery** [comment]
Closing for inactivity and labelling "future", please re-open if you plan to work on this.
